### ISSUE TITLE
[ossia-max] feature request: test to measure "time to interactive"

### ISSUE THREAD

**Issue by evanmtp** [ossia/libossia#776]
I've been testing out various nightlies of ossia-max over the past while, and sometimes there seems to be some noticeable fluctuation in how long ossia takes to load for the first time. For example, as of [f0ab85d](https://github.com/ossia/libossia/commit/f0ab85dd0451069d52e3507acf49db6f63a70a36), the first time I add any ossia-max object (ossia.parameter, ossia.remote, etc.) to a patcher, Max hangs for about 22 seconds.

I wrote a patch that could serve as the basis for a test:

[ossia_tti_test.zip](https://github.com/ossia/libossia/files/8328519/ossia_tti_test.zip)

 It uses [thispatcher] scripting to create an [ossia.parameter], bang it as soon as it loads, and use its output to stop a [timer]. This gives a value in ms for "time to interactive" (TTI).

Does it make sense to decide on a maximum acceptable load time for ossia-max and use this test to warn if it exceeds that? Even accounting for different hardware configs (and I don't know what sort of system GitHub/the CI tools are using to run unit tests), can we decide on a value that's just too slow?

I would have built this into a proper unit test, but there's currently a bug with ossia.test-logger now - please see #775 .

**Comment by jcelerier** [comment]
hm, here it seems to load in 23 ms on windows, which seems fairly reasonable...


![ossia-loading](https://user-images.githubusercontent.com/2772730/159931575-fab2d445-b905-4f84-bce6-8a4e4bed6fe0.png)

**Comment by jcelerier** [comment]
you are really getting 22s of wait there ?

**Comment by evanmtp** [comment]
I just tested on macOS and I'm also getting ~100ms load time there. On Windows, unfortunately, it's still like this for me:

https://user-images.githubusercontent.com/5342408/160052572-e013de89-8b91-4621-9139-3c7df8c26c3c.mp4

Following your instructions in #765, I ran Max via windbg and paused execution right after creating the [ossia.parameter]. Here's the screenshot:

![ossia_network_hang_01](https://user-images.githubusercontent.com/5342408/160052753-93aaa442-f067-40f3-904e-42a47b32c702.png)

I noticed that this line was being printed in the windbg Command pane while Max was hanging:

dnssd_clientstub ConnectToServer: connect()-> No of tries: 1dnssd_clientstub ConnectToServer: connect()-> No of tries: 2dnssd_clientstub ConnectToServer: connect()-> No of tries: 3dnssd_clientstub ConnectToServer: connect()-> No of tries: 1dnssd_clientstub ConnectToServer: connect()-> No of tries: 2dnssd_clientstub ConnectToServer: connect()-> No of tries: 3

and after Max stopped hanging, these lines occured:

(491c.3ff0): C++ EH exception - code e06d7363 (first chance)
(491c.3ff0): C++ EH exception - code e06d7363 (first chance)

**Comment by jcelerier** [comment]
Thanks, that's really useful ! The dnssd thing really hints towards a
network issue (it's the "technical" name of Bonjour). I'll maje you a build
without it so that we can confirm if it's the culprit

**Comment by jcelerier** [comment]
Hummmm by any chance did you maybe not install Bonjour on windows ?
https://support.apple.com/kb/dl999?locale=en_US

**Comment by evanmtp** [comment]
Eek, apparently not - I just installed it, and the hang is gone. Good catch!

Is it possible for ossia-max to check if Bonjour is installed when it loads? It would be great if it could avoid the hang and warn the user that Bonjour is required for network communication. Probably best if this doesn't prevent Max from loading ossia objects altogether, though - there might be some people who don't want to install Bonjour but still want to use ossia-max.

**Comment by jcelerier** [comment]
> Is it possible for ossia-max to check if Bonjour is installed when it loads?

yes, that's definitely something we should do

**Comment by evanmtp** [comment]
Cool! I'm glad we got to the bottom of this specific issue. Going back to the broader topic of the thread, what do you think about having a TTI automated test, in terms of both usefulness and feasibility?

For context: at a certain point, we at TML noticed that Jamoma had started to have inordinately long load/close times, to such a degree that it became very risky to put it onstage in performance/installation contexts (e.g. having a giant transient go through a PA system when DSP locks up, followed by by 4 minutes of silence while waiting for Max to close and reopen, all in front of an audience...). This is what motivated us to go looking for other options and what led us to ossia in the first place! Having some way of making sure that load times don't get out of hand due to some regression seems very important to me.

**Comment by evanmtp** [comment]
That would be great, @avilleret !

I know that I've brought up variations of this issue in the past, but I'm currently working with @navid on a concert patch and load time has become very slow as we've added more ossia objects.  It looks like we have 2259 nodes in the patch (based on passing the output of ```ossia.explorer``` to a ```t b``` and then ```counter```). It takes upwards of three minutes for the patch to load, which is quite risky for a performance patcher.

@villeret and @jcelerier , if you would be willing to look at it, I can share the patcher later, along with its dependencies. If there's anything that can be done to mitigate this issue of load time increasing drastically as more ossia objects are added, that would be great. Also, if we're doing something in the patch that's causing this, and there's another approach that we're missing, that would be good to know.

**Comment by jcelerier** [comment]
@evanmtp yes, please do share the patch! 2250 nodes should not be slow at all
